third_party/amd/lib/TritonAMDGPUTransforms/InThreadTranspose.cpp

```diff
@@ -0,0 +1,620 @@
+        continue;
+      }
+
+      // Look backward
+      if (backwardIdx != -1) {
+        auto backwardSearch =
+            traverseCFForValueDefs(candidate->getOperand(backwardIdx));
+        if (failed(backwardSearch))
+          return failure();
+        for (auto def : backwardSearch.value()) {
+          foundNetwork.sharedMemVals.push_back(def);
+          if (Operation *op = def.getDefiningOp()) {
+            // additional check, to ignore control flow operations
+            if (isa<ttg::MemDescSubviewOp, ttg::LocalAllocOp>(op))
+              nextTraversalStep.push_back(op);
+          }
+        }
+      }
+
+      // Look forward
+      if (forwardIdx != -1) {
+        auto forwardSearch =
+            traverseCFForValueUses(candidate->getResult(forwardIdx));
+        if (failed(forwardSearch))
+          return failure();
+        foundNetwork.sharedMemVals.append(forwardSearch.value().transitiveCF);
+        nextTraversalStep.append(forwardSearch.value().ops);
+      }
+    }
+    traversalStep = std::move(nextTraversalStep);
+  }
+  return foundNetwork;
+}
+
+unsigned getDimRepeats(RankedTensorType type, int dimIdx) {
+  auto loadEnc = type.getEncoding();
+  auto blockedEnc = dyn_cast<ttg::BlockedEncodingAttr>(loadEnc);
+  if (!blockedEnc)
+    return 0;
+  auto sizePerThread = blockedEnc.getSizePerThread();
+  auto lanes = blockedEnc.getThreadsPerWarp();
+  auto warps = blockedEnc.getWarpsPerCTA();
+  auto shape = type.getShape();
+  int repeats =
+      shape[dimIdx] / (sizePerThread[dimIdx] * lanes[dimIdx] * warps[dimIdx]);
+  return std::max(1, repeats);
+}
+
+llvm::FailureOr<loadStoreLoadPatternComponents>
+matchThreadRakePattern(Value operand) {
+  auto opTensorTy = cast<RankedTensorType>(operand.getType());
+  auto opEnc = opTensorTy.getEncoding();
+  auto opDotOpEnc = dyn_cast<ttg::DotOperandEncodingAttr>(opEnc);
+  if (!opDotOpEnc)
+    return failure();
+
+  int kDimNum = opDotOpEnc.getOpIdx() == 0 ? 1 : 0;
+  // TODO: support wmma
+  if (!isa<ttg::AMDMfmaEncodingAttr, ttg::AMDWmmaEncodingAttr>(
+          opDotOpEnc.getParent())) {
+    LDBG("Operand's parent encoding is not MFMA");
+    return failure();
+  }
+
+  // Find nearest local_load
+  auto localLoadSearch = findAllDefiningOps<ttg::LocalLoadOp>(operand);
+  if (failed(localLoadSearch)) {
+    LDBG("Failed to traverse local loads");
+    return failure();
+  }
+
+  if (localLoadSearch.value().size() == 0) {
+    LDBG("Did not find local load operation");
+    return failure();
+  }
+
+  loadStoreLoadPatternComponents pattern;
+
+  SetVector<Operation *> visited;
+  for (auto lLoad : localLoadSearch.value()) {
+    // find local_alloc, local_store, local_load and ttg.memdesc_subview
+    // operations
+    auto sharedMemSearch = findReachableSMemOps(lLoad, visited);
+    if (failed(sharedMemSearch)) {
+      LDBG("Failed to traverse shared memmory operation network");
+      return failure();
+    }
+    pattern = sharedMemSearch.value();
```

Comment:
So we just overwrite it and only keep the last result?  Not sure this is expected? If so why not only invoke only on the last one?